pkg/common/queue/chunkqueue.go

```diff
@@ -0,0 +1,313 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package queue
+
+import (
+	"sync"
+	"unsafe"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+)
+
+const (
+	defaultSizePerChunk = 1024
+)
+
+// ChunkQueue is a generic, efficient and GC-friendly queue
+type ChunkQueue[T any] struct {
+	// [head, tail] is the section of chunks in use
+	head int
+	tail int
+	size int
+
+	chunks       []*chunk[T]
+	chunkSize    int
+	chunkPool    sync.Pool
+	defaultValue T
+}
+
+func (q *ChunkQueue[T]) headChunk() *chunk[T] {
+	return q.chunks[q.head]
+}
+
+func (q *ChunkQueue[T]) tailChunk() *chunk[T] {
+	return q.chunks[q.tail]
+}
+
+// NewChunkQueue creates a new ChunkQueue
+func NewChunkQueue[T any]() *ChunkQueue[T] {
+	return NewChunkQueueLeastCapacity[T](1)
+}
+
+// NewChunkQueueLeastCapacity creates a ChunkQueue with an argument minCapacity.
+// It requests that the queue capacity be at least minCapacity. And it's similar
+// to the cap argument when making a slice using make([]T, len, cap)
+func NewChunkQueueLeastCapacity[T any](minCapacity int) *ChunkQueue[T] {
```

Comment:
Why using `minCapacity` instead of `initCapacity`?

Reply:
I also think the variable name is a bit opaque. However, since the capacity is always an integral multiple of `chunkLength`, it is just a lower bound of `initCapacity` and not an exact value.